### DISCUSSION TITLE
Easier wrap for template type.

### DISCUSSION THREAD

**Discussion by hzhangxyz** [pybind/pybind11#3753] [Ideas]
If I want to wrap a template type with several specialization, I need to wrap them seperately, although it is easy to write a template function to do it all. However, if it is supported by pybind, it would be better.

For example, a c++ template class is:
```c++
template<typename T>
struct Type
{ 
  void show(T arg)
  {
    std::cout << arg << "\n";
  }
};
```
Currently I need to wrap by the following code
```c++
template<template T>
void wrap(std::string type_name, std::string intro){
   py::class_<Type<T>>(m, type_name.c_str(), intro.c_str())
      .def("show", [](Type<T>& self, T arg) {self.show(arg); } );
}
```
And call wrap with different T one by one.

I hope we can write code like this:
```c++
py::template_class_<Type>
   .apply<std::tuple<int>, std::tuple<float>, ...>(   // apply to Type<int> and Type<float> ...
    mods, names, intros,
    [] (auto wrapped) {
        using TypeT = typename decltype(wrapped)::type;
        using T = typename decltype(wrapped)::arg_type<1>;
        wrapped.def("show", [](TypeT& self, T arg) {self.show(arg); });
    }
);
```

In fact, CxxWrap.jl (similar things to pybind11 for julia lang) is using this way. https://github.com/JuliaInterop/CxxWrap.jl#template-parametric-types

What else:

Maybe in the feature, we can write code in python like this:
```python
Type[int]().show(123)
Type[float]().show(1.23)
```
However, it seems currently python constructor cannot obtains the type argument in the square bracket.

**Comment by jiwaszki** [comment]
@hzhangxyz very interesting concept. I have prepared a little breakdown on my thoughts on that, as I encountered such problem in the past.

Let's start from the "core" of language itself. I will present this example:
```python
from typing import TypeVar, Generic

T = TypeVar('T', int, float)


class MyNumber(Generic[T]):
    def __init__(self, val: T) -> None:
        self.val: T = val

    def replace(self, val: T) -> None:
        self.val = val

    def is_negative(self) -> bool:
        return self.val < 0


super_int = MyNumber[int](6)
super_int.is_negative()
super_int.replace([0, 1, 2])  # error in mypy but not in the runtime

next_one = MyNumber[dict]({"a": 5})  # error in mypy but not in the runtime
next_one.is_negative()  # error in the runtime!
```
Here you may see that **in fact you can use types in brackets!** You can refer to [Python docs on Generics](https://docs.python.org/3/library/typing.html#typing.Generic) for more info on that. However there is one more important thing, on that same page you may find:

> Note The Python runtime does not enforce function and variable type annotations. They can be used by third party tools such as type checkers, IDEs, linters, etc.

That is why comments `# error in...` are present in code sample. Due to Python's nature there is no way to check types as other compiled languages do, including Julia's JIT (I am not an expert in this language - I am sorry if Julia's compiler is not called that way, please correct me if I am wrong). If you run the example, you will be able to do anything until runtime error occurs. MyPy errors are only printed after static check, which is done separately by a different program. Julia may be able to set types, use templates and provide different overloads based on argument types -- but Python is a different world. (PS: If there is a way to push vanilla Python that types are checked and overloads exists please share it with me! I am aware of hacks including decorators or dispatching arguments, but if there are more amazing tricks I would like to study them 🚀 )

With that background you may see what is the main problem when it comes to templated classes.
One trick is to wrap templated like you do in your example. In fact these templated classes are required to have distinguishable names! If package contains `MyNumberInt (MyNumber<int>)` and `MyNumberFloat (MyNumber<float>)` which can be later dispatched like:
```python
class MyNumber():
    def __new__(self, val):
        if isinstance(val, int):
            return MyNumberInt(val)
        # ...
```
... but this is simply a hacking solution. None of the returned object will be an original class which potential users are expecting to have. Maybe combining this idea with a composition and creating something like:
```python
class MyNumber():
    def __init__(self, val):
        if isinstance(val, int):
            self.my_number = MyNumberInt(val)
        # ...

    def any_method(n):
        self.my_number.any_method(n)
```
is a better way to keep the same class name if API for both classes are the same and only type is differencing them. This is the most you can squeeze from Python, and keep it somewhat readable, if you ask me. However, you still need to register multiple classes in your pybind code type by type -- C++ style.

Ok, but how it comes that types errors are shown when overloads do not exists? Internal pybind magic: https://github.com/pybind/pybind11/blob/9aa676d38db4b18d9952c70486412d20c0213f13/include/pybind11/pybind11.h#L657

So **in theory it could be possible** to use some "internal machinery" to force classes initalizers to utilise syntax of Python's typing generics and only then allow the user to create such class. **On the other hand it feels very unintuitive from Python's perspective to declare any type and create templated/overloaded solutions and even more... replacing the meaning of type hints to satisfy C++ bindings.** It seems like most of problems with the similar nature comes from the fundamental differences between two languages, and there is no easy way to solve them.

I would like to hear some more from pybind devs on that topic!

**Comment by hzhangxyz** [comment]
@jiwaszki  Can we use something like this to implement template like python class?
```python
class Template(type):

    def __new__(meta, name, bases, dict):
        dict["specialization_pool"] = {}
        dict["__class_getitem__"] = meta._class_getitem

        return type(name, bases, dict)

    def __class_getitem__(meta, params):
        if not isinstance(params, tuple):
            params = (params,)
        supertypes = tuple(object if isinstance(param, str) else param.stop for param in params)
        params = tuple(param if isinstance(param, str) else param.start for param in params)

        def _class_getitem(cls, args):
            if not isinstance(args, tuple):
                args = (args,)
            assert len(params) == len(args)
            for arg, supertype in zip(args, supertypes):
                assert issubclass(arg, supertype)
            if args not in cls.specialization_pool:
                subtype = type(
                    f"{cls.__name__}[{','.join(arg.__name__ for arg in args)}]",
                    (cls,),
                    {param: arg for param, arg in zip(params, args)},
                )
                cls.specialization_pool[args] = subtype
            return cls.specialization_pool[args]

        return type(f"Template[{','.join(params)}]", (meta,), {"_class_getitem": _class_getitem})


class MyNumber(metaclass=Template["T":int | float]):

    def __init__(self, val):
        assert isinstance(val, self.T)
        self.val = val

    def replace(self, val):
        assert isinstance(val, self.T)
        self.val = val

    def is_negative(self):
        return self.val < 0


super_int = MyNumber[int](6)
super_int.is_negative()
# super_int.replace([0, 1, 2])  # error in the runtime

# next_one = MyNumber[dict]({"a": 5})  # error in the runtime
```

should the specification `MyNumber[int]` be the subtype of `MyNumber` ?

**Comment by jiwaszki** [comment]
@hzhangxyz this could be one way to solve your issue. However, if you want some pybind based solution, I came up with a little "proof of concept". Your input gave me some great ideas! Note: I tried to rely on pybind itself, without any help of low-level Python C API.

Let's start with header file. Here we are preparing base for our bindings. That includes proxy classes and applying inheritance that will be very useful later on. Additional inheritance is providing us with methods from base class that will be uniform between specialisations, as well as playing a role of initial "constructor". Please notice two helper functions called `check_key` and `get_class_pytype`. Checking key routine is to determine if passed type is matching any allowed types. Getting type on the other hand is retrieving Python type from temporary object -- thus returning class instead of already made object. (I know it looks a little bit messy, need to improve on that part!)
```cpp
#include <string>
#include <iostream>

#include <pybind11/pybind11.h>

namespace py = pybind11;

template <class T>
class MyNumber
{
public:
    MyNumber(){};

    void show(const T &arg)
    {
        std::cout << arg << "\n";
    }
};

// Proxy base class
class PyNumberBase
{
};

// Proxy class for actual template
template <typename T>
class PyNumber : public MyNumber<T>, public PyNumberBase
{
};

bool check_key(py::object key, py::object obj)
{
    return key.is(py::type::of(obj));
}

template <typename T>
py::type get_class_pytype()
{
    auto tmp = T();
    py::object obj = py::cast(tmp);
    return py::type::of(obj);
}

// B - base class, "parent"
// C - templated class
// T - args...
template <typename B, template <typename T> typename C, typename T>
py::class_<C<T>> wrap(py::module m, std::string type_name)
{
    py::class_<C<T>, B> cls(m, type_name.c_str());

    cls.def(py::init<>());

    cls.def("show", [](C<T> &self, const T &arg)
            { self.show(arg); });

    return std::move(cls); // return allows us to update class later
}
```

Now we can provide bindings. First things first, wrap base class and name it properly, apply all "common" functions to it. Now `wrap` can create different specialisations, add specific methods. Maybe there would be a way to pack more arguments as you described with "apply approach".
```cpp
#include "code.hpp"

namespace py = pybind11;

PYBIND11_MODULE(mymodule, m)
{
    // Need to wrap first!
    py::class_<PyNumberBase> cls(m, "MyNumber");

    cls.def("__class_getitem__", [](py::object &key)
            {
                if (check_key(key, py::float_()))
                {
                    return get_class_pytype<PyNumber<float>>();
                }
                else if (check_key(key, py::int_()))
                {
                    return get_class_pytype<PyNumber<int>>();
                }
                else
                {
                    throw py::type_error("Type is not supported!");
                }
            });
    
    cls.def("__repr__", [](PyNumberBase &self) {
        return py::str("<class MyNumber>");
    });

    // MyNumber<float> bindings
    auto cls_float = wrap<PyNumberBase, PyNumber, float>(m, "MyFloat");

    // Using PyNumberBase as self is valid
    // Let's override __repr__ method from base class
    cls_float.def("__repr__", [](PyNumberBase &self) {
        return py::str("<class MyNumber[float]>");
    });

    // MyNumber<int> bindings
    auto cls_int = wrap<PyNumberBase, PyNumber, int>(m, "MyInt");

    // Using PyNumber<...> as self is also valid
    // Let's add new function
    cls_int.def("is_negative", [](PyNumber<int> &self, int val) {
        return val < 0;
    });
}
```

Let's see how it looks in the interpreter... and it looks great!
```
In [1]: import mymodule
   ...: 
   ...: mymodule.MyFloat().show(21.0)
   ...: mymodule.MyInt().show(37)
21
37
In [2]: mymodule.MyNumber[float]().show(21.0)
   ...: mymodule.MyNumber[int]().show(37)
21
37
In [3]: mymodule.MyNumber[int]().is_negative(-10)
Out[3]: True
In [4]: mymodule.MyNumber[dict]().show(2.7)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [4], in <cell line: 1>()
----> 1 mymodule.MyNumber[dict]().show(2.7)

TypeError: Type is not supported!
In [5]: mymodule.MyNumber[float]().show({2.7})
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [5], in <cell line: 1>()
----> 1 mymodule.MyNumber[float]().show({2.7})

TypeError: show(): incompatible function arguments. The following argument types are supported:
    1. (self: mymodule.MyFloat, arg0: float) -> None

Invoked with: <class MyNumber[float]>, {2.7}
In [6]: mymodule.MyNumber[float]().is_negative(-10)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 mymodule.MyNumber[float]().is_negative(-10)

AttributeError: 'mymodule.MyFloat' object has no attribute 'is_negative'
In [8]: print(mymodule.MyNumber[float]())
   ...: print(mymodule.MyNumber[int]())
<class MyNumber[float]>
<class MyNumber>
In [9]: print(mymodule.MyNumber)
<class 'mymodule.MyNumber'>
In [10]: print(isinstance(mymodule.MyNumber[float](), mymodule.MyNumber))
    ...: print(issubclass(mymodule.MyNumber[float], mymodule.MyNumber))
True
True
```

However there is one little thing to work on. As `MyNumber` is producing a type that inherits `__class_getitem__`, thus sequence of types is allowed. Overriding child classes `__class_getitem__` method so it throws error/do-nothing, could be a solution to this issue.
```
In [2]: print(mymodule.MyNumber[float][int]) # funny
   ...: print(mymodule.MyNumber[float][int]()) # funny
<class 'mymodule.MyInt'>
<class MyNumber>
```

Maybe I will have some free time soon to build upon that sample... 🤔
